libparsec/crates/client/src/workspace/fetch.rs

```diff
@@ -272,3 +274,131 @@ pub(super) async fn fetch_block(
+    #[error(transparent)]
+    InvalidCertificate(#[from] Box<InvalidCertificateError>),
+    #[error(transparent)]
+    InvalidManifest(#[from] Box<InvalidManifestError>),
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+}
+
+impl From<ConnectionError> for FetchVersionsRemoteManifestError {
+    fn from(value: ConnectionError) -> Self {
+        match value {
+            ConnectionError::NoResponse(_) => Self::Offline,
+            err => Self::Internal(err.into()),
+        }
+    }
+}
+
+pub(super) async fn fetch_versions_remote_workspace_manifest(
+    cmds: &AuthenticatedCmds,
+    certificates_ops: &CertificateOps,
+    realm_id: VlobID,
+    versions: &[VersionInt],
+) -> Result<Vec<FolderManifest>, FetchVersionsRemoteManifestError> {
+    let VlobVersionsData {
+        needed_common_certificate_timestamp,
+        needed_realm_certificate_timestamp,
+        items,
+    } = fetch_versions_vlob(
+        cmds,
+        realm_id,
+        versions.iter().map(|version| (realm_id, *version)),
+    )
+    .await?;
+
+    let mut manifests = Vec::with_capacity(items.len());
+    for (_, key_index, expected_author, expected_version, expected_timestamp, blob) in items {
+        let manifest = certificates_ops
+            .validate_workspace_manifest(
+                needed_realm_certificate_timestamp,
+                needed_common_certificate_timestamp,
+                realm_id,
+                key_index,
+                expected_author,
+                expected_version,
+                expected_timestamp,
+                &blob,
+            )
+            .await
+            .map_err(|err| match err {
+                CertifValidateManifestError::Offline => FetchVersionsRemoteManifestError::Offline,
+                CertifValidateManifestError::Stopped => FetchVersionsRemoteManifestError::Stopped,
+                CertifValidateManifestError::NotAllowed => {
+                    FetchVersionsRemoteManifestError::NoRealmAccess
+                }
+                CertifValidateManifestError::InvalidManifest(err) => {
+                    FetchVersionsRemoteManifestError::InvalidManifest(err)
+                }
+                CertifValidateManifestError::InvalidCertificate(err) => {
+                    FetchVersionsRemoteManifestError::InvalidCertificate(err)
+                }
+                CertifValidateManifestError::InvalidKeysBundle(err) => {
+                    FetchVersionsRemoteManifestError::InvalidKeysBundle(err)
+                }
+                CertifValidateManifestError::Internal(err) => {
+                    err.context("Cannot validate vlob").into()
+                }
+            })?;
+        manifests.push(manifest);
+    }
+
+    Ok(manifests)
+}
+
+struct VlobVersionsData {
+    needed_common_certificate_timestamp: DateTime,
+    needed_realm_certificate_timestamp: DateTime,
+    items: Vec<(VlobID, IndexInt, DeviceID, VersionInt, DateTime, Bytes)>,
```

Comment:
Maybe put that into a struct ?

Reply:
I didn't use a struct here given `VlobVersionsData` is not a public structure but something only used to allow code splitting between `fetch_versions_remote_workspace_manifest` and `fetch_versions_vlob`.

The `Vec<(VlobID, IndexInt, DeviceID, VersionInt, DateTime, Bytes)>` type (let's call it "the complex type" from now on ^^) comes from the API protocol, so we have to deal with it no matter what.

From this we have two choices:

1) deal with the complex type in `fetch_versions_remote_workspace_manifest`  (this is the current solution)
2) deal with the complex type in `fetch_versions_vlob` so that `fetch_versions_remote_workspace_manifest` can work with a better type (this is your proposed solution)

The trick is no matter what we do, we will end up with a loop dealing with the complex type (in case 1 it is to do the actual work, in case 2 it is to convert it to a better type).

So in the end I don't think there is a gain here.

(but again, if `VlobVersionsData` were a public structure, the logic would be much different and we would most likely benefit from having a cleaner type)

Reply:
At the end of the day, it's a matter of personal taste ^^

I like having named fields when possible. But a comment explain what are the field is enough.

Reply:
I've added a comment to provide a better documentation on the meaning of each field (and also why adding why there is no dedicated type in use here)

> At the end of the day, it's a matter of personal taste ^^

To make you feel better, the performance should be better when reusing the tuple (since the memory layout between the tuple and a dedicated type is likely different[^1] and hence the compiler cannot re-use the vec as-is) :trollface: 

[^1]: my guess is the layout is the same as long as the parameters are declared in the same order, but this is implicit without strong guarantee, so I wouldn't rely on that ^^